tests/app/questionnaire/test_value_source_resolver.py

```diff
@@ -0,0 +1,371 @@
+            [{"answer_id": "some-answer", "list_item_id": "item-1", "value": "Yes"}]
+        ),
+        list_item_id="item-1",
+    )
+
+    assert value_source_resolver.resolve(answer_source) == "Yes"
+
+
+def test_answer_source_with_list_item_id_but_answer_not_in_list_collector_or_repeat():
+    schema = get_mock_schema()
+    schema.answer_should_have_list_item_id = Mock(return_value=False)
+
+    value_source_resolver = get_value_source_resolver(
+        schema=schema,
+        answer_store=AnswerStore(
+            [{"answer_id": "some-answer", "list_item_id": "item-1", "value": "Yes"}]
+        ),
+        list_item_id="item-1",
+    )
+
+    assert value_source_resolver.resolve(answer_source) is None
+
+    value_source_resolver = get_value_source_resolver(
+        schema=schema,
+        answer_store=AnswerStore([{"answer_id": "some-answer", "value": "Yes"}]),
+        list_item_id="item-1",
+    )
+
+    assert value_source_resolver.resolve(answer_source) == "Yes"
+
+
+def test_answer_source_with_list_item_selector_location():
+    value_source_resolver = get_value_source_resolver(
+        answer_store=AnswerStore(
+            [
+                {
+                    "answer_id": "some-answer",
+                    "list_item_id": "item-1",
+                    "value": "Yes",
+                }
+            ]
+        ),
+        location=Location(
+            section_id="some-section", block_id="some-block", list_item_id="item-1"
+        ),
+    )
+
+    assert (
+        value_source_resolver.resolve(answer_source_list_item_selector_location)
+        == "Yes"
+    )
+
+
+def test_answer_source_with_list_item_selector_list_first_item():
+    value_source_resolver = get_value_source_resolver(
+        answer_store=AnswerStore(
+            [
+                {
+                    "answer_id": "some-answer",
+                    "list_item_id": "item-1",
+                    "value": "Yes",
+                }
+            ]
+        ),
+        list_store=ListStore([{"name": "some-list", "items": get_list_items(3)}]),
+    )
+
+    assert (
+        value_source_resolver.resolve(answer_source_list_item_selector_list_first_item)
+        == "Yes"
+    )
+
+
+@pytest.mark.parametrize("is_answer_on_path", [True, False])
+@pytest.mark.parametrize("is_inside_repeat", [True, False])
+def test_answer_source_with_routing_path_block_ids(is_answer_on_path, is_inside_repeat):
+    schema = get_mock_schema()
+    schema.get_block_for_answer_id = Mock(return_value={"id": f"some-block"})
+
+    location = Location(section_id="test-section", block_id="test-block")
+    id_prefix = "some" if is_answer_on_path else "some-other"
+    answer = Answer(answer_id=f"{id_prefix}-answer", value="Yes")
+
+    if is_inside_repeat:
+        location.list_item_id = answer.list_item_id = "item-1"
+        schema.answer_should_have_list_item_id = Mock(return_value=True)
+    else:
+        schema.answer_should_have_list_item_id = Mock(return_value=True)
```

Comment:
Technically yes but still worked since the list item id was `None` so it didn't evaluate it but declared it for completeness. I have decided to remove this, but explicitly assign location list item id to `None`.